### PR TITLE
Add examples to FML manifest; loading and validating

### PR DESCRIPTION
**Pull request by jhugman**
Relates to [ EXP-4204](https://mozilla-hub.atlassian.net/browse/EXP-4204).

According to [The Big O of Code Reviews](https://www.egorand.dev/the-big-o-of-code-reviews/), this is a O(_n_) change.

This PR loads and validates examples of good usage of the feature.

1. Add FeatureExample to intermediate_representation
2. Add frontend ExampleBlock and FeatureExample into ExampleBlock.
3. Load examples from disk
4. Tie frontend to intermediate representation
5. Validate examples as feature is validated
6. Add frontend for importing example blocks
7. Tie importing feature additions frontend to the parser

This PR can be best read in the order:

1. `frontend.rs`
2. `parser.rs`
3. `intermediate_representation.rs`
4. `defaults/validator.rs` and `error_path.rs`
5. `backends/frontend_manifest.rs`

### Pull Request checklist ###

- **Breaking changes**:  This PR follows our [breaking change policy](https://github.com/mozilla/application-services/blob/main/docs/howtos/breaking-changes.md)
  - [ ] This PR follows the breaking change policy:
     - This PR has no breaking API changes, or
     - There are corresponding PRs for our consumer applications that resolve the breaking changes and have been approved
- [ ] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry in [CHANGELOG.md](../CHANGELOG.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

[Branch builds](https://github.com/mozilla/application-services/blob/main/docs/howtos/branch-builds.md): add `[firefox-android: branch-name]` to the PR title.

### CODE REVIEW

File: components/support/nimbus-fml/src/backends/frontend_manifest.rs [hunk 1]
```diff
@@ -67,11 +67,32 @@ impl From<FeatureDef> for FeatureBody {
             variables.insert(f.name(), f.into());
         }
 
+        let mut examples = Vec::new();
+        for ex in value.examples {
+            examples.push(ex.into());
+        }
```
Comment (brennie) [inline components/support/nimbus-fml/src/backends/frontend_manifest.rs:73]:
```suggestion
        let examples: Vec<_> = value.examples.iter().map(ExampleBlock::from).collect();
```

File: components/support/nimbus-fml/src/backends/frontend_manifest.rs [hunk 2]
```diff
@@ -67,11 +67,32 @@ impl From<FeatureDef> for FeatureBody {
...
             variables,
             default: None,
             allow_coenrollment: value.allow_coenrollment,
+            examples,
+        }
+    }
+}
+
+impl From<FeatureExample> for ExampleBlock {
+    fn from(value: FeatureExample) -> Self {
```
Comment (brennie) [inline components/support/nimbus-fml/src/backends/frontend_manifest.rs:86]: s/value/example

File: components/support/nimbus-fml/src/backends/frontend_manifest.rs [hunk 3]
```diff
@@ -67,11 +67,32 @@ impl From<FeatureDef> for FeatureBody {
...
+        ExampleBlock::Inline(value.into())
+    }
+}
+
+impl From<FeatureExample> for InlineExampleBlock {
+    fn from(value: FeatureExample) -> Self {
```
Comment (brennie) [inline components/support/nimbus-fml/src/backends/frontend_manifest.rs:92]: s/value/example

File: components/support/nimbus-fml/src/frontend.rs [hunk 4]
```diff
@@ -133,7 +133,75 @@ pub(crate) struct ImportBlock {
...
+pub(crate) struct LabelledDefaultBlock {
+    #[serde(alias = "defaults")]
+    pub(crate) default: DefaultBlock,
+}
+
+// Rationalize all the different ways we can express imported
+// defaults and examples for a feature into a single way that the parser
+// can use to merge.
+impl From<FeatureAdditionChoices> for FeatureAdditions {
+    fn from(value: FeatureAdditionChoices) -> Self {
```
Comment (brennie) [inline components/support/nimbus-fml/src/frontend.rs:177]: s/value/choices

File: components/support/nimbus-fml/src/frontend.rs [hunk 5]
```diff
@@ -133,7 +133,75 @@ pub(crate) struct ImportBlock {
...
+        match value {
+            FeatureAdditionChoices::FeatureAdditions(a) => a,
+            FeatureAdditionChoices::AllDefaults(list) => {
+                let mut examples = Vec::default();
+                let mut defaults = Vec::default();
```
Comment (brennie) [inline components/support/nimbus-fml/src/frontend.rs:182]: `Vec::new()` or `vec![]`

File: components/support/nimbus-fml/src/frontend.rs [hunk 6]
```diff
@@ -133,7 +133,75 @@ pub(crate) struct ImportBlock {
...
+                }
+
+                FeatureAdditions { examples, defaults }
+            }
+        }
+    }
+}
+
+impl From<FeatureAdditions> for FeatureAdditionChoices {
+    fn from(value: FeatureAdditions) -> Self {
```
Comment (brennie) [inline components/support/nimbus-fml/src/frontend.rs:202]: s/value/additions

File: components/support/nimbus-fml/src/parser.rs [hunk 7]
```diff
@@ -179,6 +186,35 @@ impl Parser {
...
+        path: &FilePath,
+        manifest: &mut ManifestFrontEnd,
+    ) -> Result<()> {
+        for feature in manifest.features.values_mut() {
+            let as_typed = &feature.examples;
+            let mut inlined = Vec::with_capacity(as_typed.len());
+            for example in as_typed {
+                inlined.push(example.inline(&self.files, path)?);
+            }
+            feature.examples = inlined;
```
Comment (brennie) [inline components/support/nimbus-fml/src/parser.rs:200]: You can map with a result-returning function and then collect into a `Result<Vec<_>, _>` here, which will return either `Ok(vec)` or `Err(first_error)`:

```suggestion
            feature.examples = as_typed
                .iter()
                .map(|ex| ex.inline(&self.files, path))
                .collect::<Result<Vec<_>, _>()?;
```

File: components/support/nimbus-fml/src/parser.rs [hunk 8]
```diff
@@ -179,6 +186,35 @@ impl Parser {
...
+        }
+
+        for import in manifest.imports.iter_mut() {
```
Comment (brennie) [inline components/support/nimbus-fml/src/parser.rs:203]:
```suggestion
        for import in &mut manifest.imports {
```

File: components/support/nimbus-fml/src/parser.rs [hunk 9]
```diff
@@ -252,21 +288,27 @@ impl Parser {
...
+                        format!("Cannot override defaults for `{f}` feature from {child_id}"),
                     )
                 })?;
+                // FeatureAdditions holds the extra examples and defaults for this feature.
+                let additions: FeatureAdditions = feature_additions.clone().into();
+
+                // d.i) Append the imported list of examples to the original feature examples and…
+                for ex in &additions.examples {
+                    feature_def.examples.push(ex.into());
+                }
```
Comment (brennie) [inline components/support/nimbus-fml/src/parser.rs:307]:
```suggestion
                feature_def.examples.extend(additions.examples.iter());
```

File: components/support/nimbus-fml/src/parser.rs [hunk 10]
```diff
@@ -443,12 +485,11 @@ fn merge_map<T: Clone>(
...
-    for (id, defaults) in &b.features {
-        let mut defaults = defaults.clone();
+    for (id, additions) in &b.features {
         if let Some(existing) = block.features.get_mut(id) {
-            existing.append(&mut defaults);
+            existing.merge(additions);
         } else {
-            block.features.insert(id.clone(), defaults.clone());
+            block.features.insert(id.clone(), additions.clone());
         }
```
Comment (brennie) [inline components/support/nimbus-fml/src/parser.rs:493]: This should use the Entry API, which avoids double-hashing of the key in the empty case.

```rust
block.features
    .entry(id.clone())
    .and_modify(|existing| existing.merge(additions))
    .or_insert_with(|| additions.clone());
```

File: components/support/nimbus-fml/src/parser.rs [hunk 11]
```diff
@@ -491,6 +532,54 @@ fn check_can_import_list(
...
+        };
+    }
+}
+
+impl FeatureAdditions {
+    fn inline(&self, files: &FileLoader, root: &FilePath) -> Result<Self> {
+        let mut examples = Vec::default();
+        for ex in &self.examples {
+            examples.push(ex.inline(files, root)?);
+        }
```
Comment (brennie) [inline components/support/nimbus-fml/src/parser.rs:570]:
```suggestion
        let mut examples: Vec<_> = self.examples.iter().map(|ex| ex.inline(files, root)).collect::<Result<_, _>>()?;
```

File: components/support/nimbus-fml/src/parser.rs [hunk 12]
```diff
@@ -491,6 +532,54 @@ fn check_can_import_list(
...
+            }
+        })
+    }
+}
+
+impl FeatureAdditionChoices {
+    fn merge(&mut self, other: &Self) {
+        match (self, other) {
+            (Self::FeatureAdditions(a), Self::FeatureAdditions(b)) => a.merge(b),
+            _ => unreachable!("FeatureAdditionChoices should have been rationlized already. This is a bug in nimbus-fml"),
```
Comment (brennie) [inline components/support/nimbus-fml/src/parser.rs:560]: typo: rationlized -> rationalized.

File: components/support/nimbus-fml/src/parser.rs (hunk 10 above)
Comment (jhugman) [reply components/support/nimbus-fml/src/parser.rs:493]: Nice!

File: components/support/nimbus-fml/src/parser.rs (hunk 11 above)
Comment (jhugman) [reply components/support/nimbus-fml/src/parser.rs:570]: Nice.

File: components/support/nimbus-fml/src/parser.rs (hunk 7 above)
Comment (jhugman) [reply components/support/nimbus-fml/src/parser.rs:200]: Neat!

File: components/support/nimbus-fml/src/frontend.rs [hunk 13]
```diff
@@ -323,11 +395,14 @@ impl ManifestFrontEnd {
             for (fnm, field) in &body.variables {
                 fields.push(self.get_prop_def_from_feature_field(fnm, field));
             }
+            let examples = body.examples.iter().map(|ex| ex.into()).collect();
```
Comment (brennie) [inline components/support/nimbus-fml/src/frontend.rs:398]: Into::into should work here instead of making a closure.

**Review by brennie — APPROVED** [review]

**Pull request merged**